### PR TITLE
server: Redirect fallback browsers to /browsehappy

### PR DESCRIPTION
**Pull request by scinos**
#### Background

See #53222

#### Changes proposed in this Pull Request

* Drop support for fallback browsers (eg: IE 11). When those browsers hit Calypso they will get redirected to `/browsehappy`.
* Add a link to `/browsehappy` page to allow users to continue loading the potentially broken page anyway. This is stored in a cookie which expires after 24 hours.
* This will only apply to browsers in dev, horizon, and wpcalypso environments to start.

Screen Shot 2021-07-30 at 10 55 14 AM

#### Testing instructions

(You can find a link to calypso.live in this comment: https://github.com/Automattic/wp-calypso/pull/53221#issuecomment-886468659)

Follow this steps with an old browser (i.e. IE11 on browserstack)

##### Redirect flow
* Load the live branch -> It should redirect to `/browsehappy` and show a link to continue loading the page.
* If you click that link -> the original URL will be loaded.

##### Reload flow
* Do the 'Redirect flow'
* Refresh the page -> the original URL is loaded, it doesn't redirect to`/browsehappy`

You can also test the page design by visiting `/browsehappy`.

### CODE REVIEW

**Review by tyxla — COMMENTED** [review]
This is exciting to see! 🎉 

Also, it tests as explained in the PR.

However, a bunch of tests seem to be failing. Is it because we need to instruct `browserslist-useragent` to work with evergreen? Or perhaps there are some `matchesUA` mocks bleeding between tests?

**Comment by scinos** [comment]
> However, a bunch of tests seem to be failing. Is it because we need to instruct browserslist-useragent to work with evergreen? Or perhaps there are some matchesUA mocks bleeding between tests?

They were failing because the test mocks a non-evergreen browser by default. I added a bunch of `withEvergreenBrowser()` and `bypassTargetRedirection=true` when we want to test the fallback behaviour.

**Comment by scinos** [comment]
My plan is to have `bypassTargetRedirection=true` working for a few days in case something huge breaks so HE has that tool to help the users. If everything goes well, I'll delete the support for `bypassTargetRedirection` which includes deleting support for `fallback` entirely.

**Review by tyxla — APPROVED** [review]
LGTM 👍

**Comment by scinos** [comment]
Update: added a link to the /browsehappy page to continue loading the original url. This preference is stored in a session cookie.

File: client/server/pages/test/index.js
Comment (tyxla) [inline client/server/pages/test/index.js:1222]: Could you please elaborate on this one?

**Review by tyxla — APPROVED** [review]
Confirmed this works great with the cookie. Nice work!

As a side note, this page would be a great candidate for localization, cc @Automattic/i18n

File: client/server/pages/test/index.js
Comment (scinos) [reply client/server/pages/test/index.js:1222]: It is a small drive-by improvement.

This tests is asserting the behaviour when `config.isEnabled( 'jetpack-cloud/connect' )` is false (https://github.com/Automattic/wp-calypso/blob/trunk/client/server/pages/index.js#L627), but it was relying on the fact that the mocked `config.isEnabled()` returns `undefined` by default. I think being explicit about which config key is expected to be disabled makes the test more correct.

Comment (tyxla) [reply client/server/pages/test/index.js:1222]: Good one 👍

File: client/document/browsehappy.jsx
Comment (sgomes) [inline client/document/browsehappy.jsx]: It looks like no translation was being done on this page before, but now that it's going to be shown to more people it's probably important to make sure it's translated.

File: client/document/browsehappy.jsx
Comment (sgomes) [inline client/document/browsehappy.jsx]: "cannot be used" seems a bit hash, and potentially misleading. How about "may not work correctly with" instead?

File: client/document/browsehappy.jsx
Comment (sgomes) [inline client/document/browsehappy.jsx]: "upgrade" has the connotation that the suggested browsers on this list are better than the one the user is currently using. Which may well be a modern browser, just one we don't explicitly support.

How about "switch to a supported browser"?

**Review by sgomes — COMMENTED** [review]
Added some comments regarding the text in the `/browsehappy` page.

File: client/server/pages/index.js
Comment (jsnajdr) [inline client/server/pages/index.js:625]: This change causes `res.redirect()` to be called and then `next()`, too. Before this patch, `next()` was not called in case of redirect, interrupting the handler chain. Was this an intent?

File: client/components/empty-content/index.jsx
Comment (jsnajdr) [inline client/components/empty-content/index.jsx]: This can just be a `PropTypes.node`. Addressing in #54790 together with 20+ other components where this prop type is also unnecessarily complicated.

File: client/server/pages/index.js
Comment (jsnajdr) [reply client/server/pages/index.js:625]: Oh, I just noticed I read the patch backwards 🙂 It's fixing the exact bug I'm describing, not introducing it!

**Comment by jsnajdr** [comment]
I reworked this PR to implement the new design suggested by @olaolusoga in p4TIVU-9Ir-p2. This is how it looks now:

Screenshot 2021-07-26 at 10 02 51

One thing I'm missing is an SVG file for the new Red Unplugged Socket graphics that would replace the current Blue Sinking Dinosaur. @olaolusoga Do you have an SVG file for that at hand?

The "view unsupported browsers" link points to https://wordpress.com/support/browser-issues/#supported-browsers. If there is a better destination for that, I'm ready to replace it.

If the `/browsehappy` page was redirected to from an attempt to load a Calypso page, i.e, when it has a `from=url` query param, it will show a "continue loading the page anyway" link. Following it will load the Calypso page anyway and will set a cookie that disables the unsupported browser detection for 24 hours. That's mainly for false positives where we detect a browser as unsupported even if it would be able to load Calypso just fine. Such mistakes always happen and it's a good idea to provide an escape hatch. The copy of the link is up for review.

There are no longer any links to wp-admin dashboard of your primary blog, so I was able to remove some code that was determining the primary blog URL.

To make JS-imported CSS styles (`import './style.scss'`) work on the server-side-rendered `/browsehappy` page, I created a new `entry-browsehappy` entrypoint. The rendered page then loads the webpack-generated CSS styles for that entrypoint. And it doesn't bother to load the entrypoint's JS assets, even though they were generated, because the page is 100% static with no JS/React functionality.

**How to test:**
The `/browsehappy` redirect is done only when the `redirect-fallback-browsers` feature flag is enabled. In this PR I'm enabling it for `development`, `horizon` and `wpcalypso`. It remains disabled for vast majority of users on `stage` and `production`.

If you load a Calypso page (e.g., `/read`) with an unsupported browser, it should redirect to the `/browsehappy` page.

That redirect can be disabled by a `bypassTargetRedirection=true` query arg or with a `bypass_target_redirection=true` cookie.

The presence of the `bypassTargetRedirection=true` query arg will set the cookie for 24 hours, so the disable persists across navigations and reloads for some time.

The "continue anyway" link on the `/browsehappy` page should navigate back to the original page, with the `bypassTargetRedirection=true` query arg appended. That ensures we are not redirected again.

**Comment by sgomes** [comment]
I still have reservations about the copy in the updated design. It will be shown both for old browsers that don't work correctly, in which case the copy is fine, but as far as I understand it will also be shown in modern browsers that just happen to be unsupported, such as Samsung Internet.

For those browsers, it seems incorrect to be saying "this page cannot be viewed by our browser". Can we please change this to "this page may not work correctly in your browser"?

**Comment by jsnajdr** [comment]
> as far as I understand it will also be shown in modern browsers that just happen to be unsupported, such as Samsung Internet.

To clarify: such a browser is merely not detected by our UA check, but is actually able to execute the `evergreen` app, with all the latest untranspiled JS syntax?

Are there known browsers that we know can load Calypso, but are not detected as `evergreen`? And why can't we detect them as such?

**Comment by sgomes** [comment]
> To clarify: such a browser is merely not detected by our UA check, but is actually able to execute the `evergreen` app, with all the latest untranspiled JS syntax?

That is my expectation, yes. Samsung Internet is Chromium-based and regularly updated, so I don't see any reason to think otherwise.

I don't know a good way of testing it, however, since we have a flag to force fallback but not one to force evergreen. Perhaps we could point Browserstack to a local, modified build?

> Are there known browsers that we know can load Calypso, but are not detected as `evergreen`? And why can't we detect them as such?

Anything Chromium-based **which is frequently updated** and isn't Chrome or Edge should fit the bill.

As for why we can't detect them, nothing's really stopping us. We could certainly have a list of browsers we don't officially support but recognise and don't show this prompt for, perhaps? Or we could add them to the list of officially supported browsers and make sure we test on them regularly.

But not everything that's Chromium-based will work correctly, however. Some browsers like UC may be Chromium-based but tend to use very old builds of Chromium, and have a number of features removed.

Ideally we shouldn't be using UAs at all, but rather running some sort of feature detection. That's difficult to do with language features, however, as syntax errors at the parser level essentially make a script unrecoverable. We'd need to use `eval` or other unpleasant hacks.

**Comment by olaolusoga** [comment]
@jsnajdr  Here's an SVG of the illustration https://www.dropbox.com/t/locN3yF0IbZIZOEd

**Comment by olaolusoga** [comment]
"Unfortunately this page may not work correctly in your browser" sounds fine, since we give them the option to view anyways.

**Comment by jsnajdr** [comment]
I updated the illustration and the copy and now the page looks like this:

Screenshot 2021-07-27 at 10 56 05

The PR is ready for final review.

> we have a flag to force fallback but not one to force evergreen.

@sgomes Adding such a flag shouldn't be hard and it would be useful for testing.

About false positives and false negatives when detecting the browser: what if we displayed the `/browsehappy` page _only_ for Internet Explorer? That alone catches a vast majority of actually deprecated browsers. And it prevents more false positives, where we mark a good browser as deprecated, at the cost of more false negatives, where we fail to detect a deprecated browser and display a broken app. Is that a good tradeoff? Which failure mode is worse?

**Comment by scinos** [comment]
> About false positives and false negatives when detecting the browser: what if we displayed the `/browsehappy` page only for Internet Explorer?

Isn't that cover already by the "Continue loading" link and the updated copy? IMO the "Continue loading" is more than enough: the error page will be correct in most cases (Calypso will actually break in that browser), and for the cases we miss, we give the user a workaround that doesn't require mess with UA strings or contacting HE.

If we get reports that a particular browser is constantly misidentified as "broken" and it has support by [Browserslist](https://github.com/browserslist/browserslist#browsers) we can evaluate the impact of adding it to the list of allowed browsers.

**Comment by jsnajdr** [comment]
> Isn't that cover already by the "Continue loading" link and the updated copy?

Yes, it's covered, but the user still gets the banner every 24 hours. We can certainly iterate after merging this one.

**Review by noahtallen — COMMENTED** [review]
The code looks solid to me. It looks like there was a dependency issue in the e2e build, which prevented the link from showing up, and `http://calypso.localhost:3000?forceFallback=true` unfortunately didn't work for me. Should that work in dev mode?

edit: I noticed it is not supposed to work in dev mode :)

**Comment by scinos** [comment]
E2E are failing because `fallback` fails to compile when copying modules. Log:

```
#15 15.43 $ node bin/copy-production-modules.js
16:00:20   #15 15.49 [evergreen] <s> [webpack.Progress] 15% building 0/5 entries 5097/5167 dependencies 460/2200 modules
16:00:20   #15 15.58 Error: package not found: react-native, dependency of ./node_modules/@emotion/native
16:00:20   #15 15.58     at processPackage (/calypso/bin/copy-production-modules.js:99:10)
16:00:20   #15 15.58     at processPackage (/calypso/bin/copy-production-modules.js:109:3)
16:00:20   #15 15.58     at processPackages (/calypso/bin/copy-production-modules.js:38:3)
16:00:20   #15 15.58     at Object.<anonymous> (/calypso/bin/copy-production-modules.js:21:26)
16:00:20   #15 15.58     at Module._compile (internal/modules/cjs/loader.js:1085:14)
16:00:20   #15 15.58     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
16:00:20   #15 15.58     at Module.load (internal/modules/cjs/loader.js:950:32)
16:00:20   #15 15.58     at Function.Module._load (internal/modules/cjs/loader.js:790:14)
16:00:20   #15 15.58     at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
16:00:20   #15 15.58     at internal/main/run_main_module.js:17:47
16:00:20   #15 15.59 error Command failed with exit code 1.
```

**Comment by noahtallen** [comment]
> E2E are failing because fallback fails to compile when copying modules. Log:

For some reason, `yarn build` succeeds for me locally, even after deleting `node_modules` and reinstalling. Interestingly, on my system, `node_modules/@emotion/native` does exist, but `node_modules/react-native` does not exist.

**Comment by noahtallen** [comment]
I've rebased to resolve merge conflicts and to see if it will have an impact on the build

Note: lint issues will be fixed once #55033 is merged and rebased here.

Update: lint issues are resolved.

**Comment by noahtallen** [comment]
I'm able to replicate the build error with `CALYPSO_ENV='production' yarn build`. Interestingly, according to yarn.lock, react-native is not a dependency of @emotion/native. It is listed as a peer dependency in @emotion/native's package.json, though.

In other words, yarn correctly (I think) does not install react-native, but copy-production-modules expects all peer dependencies to exist:

https://github.com/Automattic/wp-calypso/blob/47f6057f24cf4f976e0baba2d541f8f08eb3231e/bin/copy-production-modules.js#L79

The build succeeds if I delete `peerDependencies` from that array, but I imagine that could have other unintended consequences. What if we made peer dependencies optional? E.g. we copy them if they exist in node_modules, but don't fail if they don't exist.

Edit: implemented in 2508635! Let me know if that was the right approach. 

Edit 2: After discussion, I removed that commit and instead added an anchor with the right classnames, so that the style is the same, but without importing the package which causes the server build to fail. (see c15133e) This seems like a fair workaround until more recent versions of wordpress components are merged. See

File: client/webpack.config.js
Comment (noahtallen) [inline client/webpack.config.js:158]: Out of curiosity, why is this `landing`? What's the difference between that and `document`?

File: client/landing/browsehappy/index.jsx
Comment (noahtallen) [inline client/landing/browsehappy/index.jsx]: The server crashes when `from` doesn't exist in the query args (when accessing the page directly)

Comment (noahtallen) [reply client/landing/browsehappy/index.jsx]: Resolved in 0e03a68

**Comment by noahtallen** [comment]
Current testing status:

- ✅ Redirects to browsehappy when using IE11
- ✅ Does not redirect after going through the "continue anyways" flow when using IE11
- ✅ Loads browsehappy when accessing the URL directly (with no query params)
- ✅ Redirects to browsehappy when `?forceFallback=true` is added to the URL when using a supported browser (I was dev mode, which always disables the fallback check. This works in the calypso.live build.)

**Comment by scinos** [comment]
Peer dependencies shouldn't be optional, packages may break if peer dependencies are not found.

The solution is, IMO, ensure that all peerDependencies requirements are meet. In other words, add react-native to every consumer of @emotion/native (assuming we really need @emotion/native, of course)

**Comment by noahtallen** [comment]
We don't actually require @emotion/native, it's a dependency of @wordpress/components. Do we add react native everywhere we add @wordpress/components? That seems like overkill since we don't use react native anywhere 🤔 

What if we were smarter about it and didn't crash when the "missing" dep came from a transitive dependency?

File: client/server/pages/index.js
Comment (noahtallen) [inline client/server/pages/index.js:151]: What's the purpose of the calypso env check? I noticed that `?forceFallback=true` does not work in dev mode -- it might be helpful for testing the unsupported browser page, not sure if it has an impact otherwise.

File: bin/copy-production-modules.js
Comment (noahtallen) [inline bin/copy-production-modules.js]: Circling back, the real problem is that transitive dependencies may or may not have a "correct" package.json layout for peer/optional deps. We aren't really in control of that, and IMO we're lucky that in this case, we could submit an upstream patch to Gutenberg very easily if we needed to. Since the rest of the build chain and entire app and test suite works fine, does it really make sense to fail this part of the server build?

At any rate, I've double checked that the latest version of @wordpress/components (v15) does not include @emotion/native, so once that is upgraded across all of wp-calypso, we should be ok here. Though, I did think the wp monorepo update was blocked until we deprecated IE11 :P Do we need a temporary fix? We could just add react-native to the root package.json temporarily, then it would exist in the directory to be found. We could also create a list of packages we don't care about not existing here.

Comment (fullofcaffeine) [reply bin/copy-production-modules.js]: > Circling back, the real problem is that transitive dependencies may or may not have a "correct" package.json layout for peer/optional deps. 

Could you elaborate on that? 🤔 I can see that `emotion/native` does include `react-native` as a `peerDependency`: https://github.com/emotion-js/emotion/blob/main/packages/native/package.json#L29. Or were you referring to another dependency?

Comment (fullofcaffeine) [reply bin/copy-production-modules.js]: Ah 💡, I re-read and now I understood. The transitive dependency in question here is `@wordpress/components`, which doesn't include `react-native` as a `peerDependency`, even though it does seem to consume it directly and through other libraries (like `emotion`).

Comment (fullofcaffeine) [reply bin/copy-production-modules.js]: I'm a bit confused about considering peer dependencies as optional here, though. Wouldn't that potentially cause other issues at runtime? Isn't what @scinos said [here](https://github.com/Automattic/wp-calypso/pull/53221#issuecomment-889474797) still relevant?

**Comment by scinos** [comment]
> We don't actually require @emotion/native, it's a dependency of @wordpress/components. Do we add react native everywhere we add @wordpress/components? That seems like overkill since we don't use react native anywhere 🤔

But that's the requirement of a library we have chosen to trust. `@emotion/native` is saying "everybody that uses me needs to have `react-native` installed. `@wordpress/components` made a strange choice: they don't provide `@emotion/native` nor pass that requirement to the consumer (i.e. us). To be fair, in NPM land there was no consensus about the transitivity of peer dependencies (afaik), but Yarn made it very explicit: you either have to fulfil all `peerDependencies` of your dependencies, or declare them as `peerDependencies` yourself. 

`peerDependencies` can't be treated as optional. If the authoritative source (the package's `package.json`) says it needs a dependency, we should trust it. Even if that dependency is not actually used today in the codepath we depend on, there is no guarantee that it will remain the case in the next package update.

I think the solution is to add `react-native` as a dependency. Yarn is becoming more strict about missing peer dependencies and I had to add it to my yarn v3 update branch (among many others) anyway. For more discussion in this topic, see https://github.com/Automattic/wp-calypso/pull/51719/files#r666201334

My preference would be to add `react-native` to each `package/*/package.json` that has an import that requires `react-native` (if any), and add `react-native` as a real dependency to any app (`app/*/package.json` and `client/package.json`) that has a component that requires `react-native` either directly or transitively. Given that it will (or should) be a short-lived hack until yarn-v3 or `@wordpress/*` lands, it will be ok to add `react-native` to the root's `package.json` as a "catch all" too.

Alternative: why do we have this problem right now? We already have a server app with a bunch of SSRd pages that don't require `react-native`. Is there something in the new `/browserhappy` landing page that we can change to avoid bringing in a dependency that needs `react-native`?

**Comment by noahtallen** [comment]
> Alternative: why do we have this problem right now? We already have a server app with a bunch of SSRd pages that don't require react-native. Is there something in the new /browserhappy landing page that we can change to avoid bringing in a dependency that needs react-native?

The new design imports from @wordpress/components. I think @wordpress/components must not have been a dependency of anything in the server build previously:

https://github.com/Automattic/wp-calypso/blob/0e03a6842bf741e7c0afaa450222de468a48e96f/client/landing/browsehappy/index.jsx#L9

From what I can tell, that's the only import, so we could use a different component there in the meantime. I imagine that would be easier than adding react-native everywhere, especially since this is (hopefully) a short term problem.

> peerDependencies can't be treated as optional. If the authoritative source (the package's package.json) says it needs a dependency, we should trust it.

Thanks for expanding on that! I guess a stricter approach will be good for the ecosystem in the long run, but it is definitely painful now. :) And, with yarn 3, we won't get into a position where a small part of the build fails, because it should hopefully be failing at install time.

File: bin/copy-production-modules.js
Comment (noahtallen) [reply bin/copy-production-modules.js]: In this case, it doesn't cause issues at runtime, because the react-native portions of @wordpress/components are irrelevant for calypso. We've been using this version of @wordpress/components for many months without issues, and with the missing peer dependency, and it doesn't cause problems elsewhere in the build. It only shows up now because this is the only part of our build chain which crashes if a peer dependency doesn't exist for a transitive dependency, and this is the first time we include @wordpress/components in the server bundle

**Comment by noahtallen** [comment]
> we could use a different component there in the meantime

It looks like we get the same style with `<a class="components-button is-primary"` without the failing build, so I think that's a fair workaround for now!

**Comment by noahtallen** [comment]
I think this is once again ready for final testing. You can find links to calypso.live in this comment: https://github.com/Automattic/wp-calypso/pull/53221#issuecomment-886468659

And you can use a free browserstack account to quickly test redirect behavior in IE 11 :) 

You can also test the page directly by navigating to `/browsehappy`

**Review by fullofcaffeine — APPROVED** [review]
This tests well!

**Comment by fullofcaffeine** [comment]
We'll also need to update the docs here and update the unsupported message to say IE 11 or earlier:

Screenshot from 2021-07-30 14-16-13

Does anyone know where I can edit that?

File: client/server/pages/index.js
Comment (jsnajdr) [reply client/server/pages/index.js:151]: In dev mode, we don't build the `fallback` target. And we don't even write the compiled files to any directory on disk. Webpack runs inside the Node.js server, compiling and recompiling as the source files change, and serves the JS and CSS assets from memory using `webpack-dev-middleware` plugged into the Express.js server.

File: client/webpack.config.js
Comment (jsnajdr) [reply client/webpack.config.js:158]: The component in `client/document` is ever ran only in the server, rendering the HTML document for the page. The component in `client/landing`, however, is part of the webpack client-side bundle for the entrypoint and runs mainly on the client. It runs on the server only when the page uses SSR.

`landing` is a directory for landing pages that have a separate entrypoint and are not part of the `entry-main` bundle.

File: client/landing/browsehappy/index.jsx
Comment (jsnajdr) [inline client/landing/browsehappy/index.jsx:29]: Using `@wordpress/components` for SSR will be unfortunately harder than this. When I tried to use the latest `@wordpress/components` in a toy Next.js app, I got this error:

```
ReferenceError: CSS is not defined
    at space (/Users/jsnajdr/src/nextjs-blog/node_modules/@wordpress/components/build/ui/utils/space.js:38:3)
    at Object.<anonymous> (/Users/jsnajdr/src/nextjs-blog/node_modules/@wordpress/components/build/utils/config-values.js:72:42)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:14)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/Users/jsnajdr/src/nextjs-blog/node_modules/@wordpress/components/build/utils/style-mixins.js:61:44)
    at Module._compile (internal/modules/cjs/loader.js:1085:14) {
```
There is a `space()` helper function that calls `CSS.supports()`, which is available only in a browser environment. FYI @sarayourfriend

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6317691

Thank you @scinos for including a screenshot in the description! This is really helpful for our translators.

File: client/landing/browsehappy/index.jsx
Comment (noahtallen) [reply client/landing/browsehappy/index.jsx:29]: That's too bad!

File: client/webpack.config.js
Comment (noahtallen) [reply client/webpack.config.js:158]: Thanks for explaining!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6317691).

**Pull request merged**
